qurro/q2/plugin_setup.py

```diff
@@ -101,3 +106,49 @@
     name=short_desc.format("loading"),
     description=long_desc.format("loading"),
 )
+
+qarcoal_params = {
+    "num_string": Str,
+    "denom_string": Str,
+    "taxonomy": Metadata,
+    "samples_to_use": Metadata,
+    "allow_shared_features": Bool,
+}
+
+qarcoal_param_descs = {
+    "num_string": "numerator string to search for in taxonomy",
+    "denom_string": "denominator string to search for in taxonomy",
+    "taxonomy": "Qiime2 Metadata with taxonomy information",
+    "samples_to_use": "Qiime2 Metadata with samples to use",
+    "allow_shared_features": (
+        "whether to raise error if features \
+        are shared between numerator and denominator"
+    ),
+}
+
+plugin.methods.register_function(
+    function=qarcoal,
+    inputs={"table": FeatureTable[Frequency]},
```

Comment:
This makes sense. If I remember correctly I tried to do that and I think it told me that Q2-Metadata wasn't a valid input (but rather a parameter). Since we're changing taxonomy to an Artifact it should work.

Reply:
Understood. It makes sense that Q2 would handle Metadata that way... yeah, using a more specific type for taxonomy should make it usable as an input.

Reply:
Ok, so (possible) hiccup? When I try to tell `plugin_setup.py` to take `FeatureData[TSVTaxonomyFormat]` as an *input*, it yells at me: `no such option: --m-taxonomy-file`. I think the way Q2 is set up, feature data can only be input as a parameter rather than an input. Is there a plugin that explicitly takes a `FeatureData` object as an input rather than a parameter?

Reply:
That's super weird, huh. I know q2-taxa (e.g. the barplot visualizer) has a bunch of stuff with `--i-taxonomy` inputs -- see [here](https://github.com/qiime2/q2-taxa/blob/master/q2_taxa/plugin_setup.py). Although it looks like they just take in `FeatureData[Taxonomy]`, rather than `FeatureData[TSVTaxonomyFormat]`. Maybe that is what makes the difference...? (I don't know why it would.)

...Actually, since it seems like `FeatureData[Taxonomy]` is [guaranteed](https://github.com/qiime2/q2-types/blob/master/q2_types/feature_data/_format.py#L19-L43) to have `Feature ID` and `Taxon` columns, it should still be usable here. See [here](https://github.com/qiime2/q2-taxa/blob/master/q2_taxa/_visualizer.py) for an example of how the barplot visualizer accepts the taxonomy input -- looks like they just load it as a pandas Series.